File: golang/columns/validator_test.go
package columns

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"

	flyql "github.com/iamtelescope/flyql/golang"
	"github.com/iamtelescope/flyql/golang/flyqltype"
	"github.com/iamtelescope/flyql/golang/transformers"
)

func validatorDataPath() string {
	_, file, _, _ := runtime.Caller(0)
	return filepath.Join(filepath.Dir(file), "..", "..", "tests-data", "columns", "validator.json")
}

type validatorExpectedDiag struct {
	Code            string `json:"code"`
	Severity        string `json:"severity"`
	Range           [2]int `json:"range"`
	MessageContains string `json:"message_contains"`
}

type validatorColumnDef struct {
	Name     string                        `json:"name"`
	Type     string                        `json:"type"`
	Children map[string]validatorColumnDef `json:"children,omitempty"`
}

type validatorTestCase struct {
	Name                string                  `json:"name"`
	Input               string                  `json:"input"`
	Capabilities        *testCapabilities       `json:"capabilities,omitempty"`
	Columns             []validatorColumnDef    `json:"columns"`
	ExpectedDiagnostics []validatorExpectedDiag `json:"expected_diagnostics"`
}

type validatorTestSuite struct {
	Tests []validatorTestCase `json:"tests"`
}

func makeValidatorSchema(defs []validatorColumnDef) *flyql.ColumnSchema {
	m := make(map[string]*flyql.Column, len(defs))
	for _, d := range defs {
		col := buildColumn(d)
		m[d.Name] = &col
	}
	return flyql.NewColumnSchema(m)
}

func buildColumn(d validatorColumnDef) flyql.Column {
	t, _ := flyql.ParseType(d.Type)
	col := flyql.NewColumn(d.Name, t)
	if len(d.Children) > 0 {
		col.Children = make(map[string]*flyql.Column, len(d.Children))
		for childName, childDef := range d.Children {
			child := buildColumn(childDef)
			col.Children[childName] = &child
		}
	}
	return col
}

func TestSharedValidatorFixtures(t *testing.T) {
	data, err := os.ReadFile(validatorDataPath())
	if err != nil {
		t.Fatalf("Cannot read fixture file: %v", err)
	}
	var suite validatorTestSuite
	if err := json.Unmarshal(data, &suite); err != nil {
		t.Fatalf("Cannot unmarshal fixture: %v", err)
	}

	for _, tc := range suite.Tests {
		t.Run(tc.Name, func(t *testing.T) {
			caps := Capabilities{Transformers: true}
			if tc.Capabilities != nil && tc.Capabilities.Transformers != nil {
				caps.Transformers = *tc.Capabilities.Transformers
			}
			parsed, parseErr := Parse(tc.Input, caps)
			if parseErr != nil {
				parsed = nil
			}
			schema := makeValidatorSchema(tc.Columns)
			diags := Diagnose(parsed, schema, nil)
			if diags == nil {
				diags = []flyql.Diagnostic{}
			}
			if len(diags) != len(tc.ExpectedDiagnostics) {
				t.Fatalf("Expected %d diagnostics, got %d", len(tc.ExpectedDiagnostics), len(diags))
			}
			for i, expected := range tc.ExpectedDiagnostics {
				if diags[i].Code != expected.Code {
					t.Errorf("diag[%d].Code = %q, want %q", i, diags[i].Code, expected.Code)
				}
				if string(diags[i].Severity) != expected.Severity {
					t.Errorf("diag[%d].Severity = %q, want %q", i, diags[i].Severity, expected.Severity)
				}
				if diags[i].Range.Start != expected.Range[0] || diags[i].Range.End != expected.Range[1] {
					t.Errorf("diag[%d].Range = {%d, %d}, want {%d, %d}",
						i, diags[i].Range.Start, diags[i].Range.End, expected.Range[0], expected.Range[1])
				}
			}
		})
	}
}

func TestDiagnose_UnknownColumn(t *testing.T) {
	parsed, err := Parse("foo", Capabilities{Transformers: true})
	if err != nil {
		t.Fatal(err)
	}
	columns := []flyql.Column{flyql.NewColumn("level", flyql.TypeString)}
	diags := Diagnose(parsed, flyql.FromColumns(columns), nil)
	if len(diags) != 1 {
		t.Fatalf("Expected 1 diagnostic, got %d", len(diags))
	}
	if diags[0].Code != flyql.CodeUnknownColumn {
		t.Errorf("Code = %q, want %q", diags[0].Code, flyql.CodeUnknownColumn)
	}
	if diags[0].Range.Start != 0 || diags[0].Range.End != 3 {
		t.Errorf("Range = {%d, %d}, want {0, 3}", diags[0].Range.Start, diags[0].Range.End)
	}
}

func TestDiagnose_UnknownTransformer(t *testing.T) {
	parsed, err := Parse("level|zzzz", Capabilities{Transformers: true})
	if err != nil {
		t.Fatal(err)
	}
	columns := []flyql.Column{flyql.NewColumn("level", flyql.TypeString)}
	diags := Diagnose(parsed, flyql.FromColumns(columns), nil)
	if len(diags) != 1 {
		t.Fatalf("Expected 1 diagnostic, got %d", len(diags))
	}
	if diags[0].Code != flyql.CodeUnknownTransformer {
		t.Errorf("Code = %q, want %q", diags[0].Code, flyql.CodeUnknownTransformer)
	}
	if diags[0].Range.Start != 6 || diags[0].Range.End != 10 {
		t.Errorf("Range = {%d, %d}, want {6, 10}", diags[0].Range.Start, diags[0].Range.End)
	}
}

func TestDiagnose_ValidColumn(t *testing.T) {
	parsed, err := Parse("level", Capabilities{Transformers: true})
	if err != nil {
		t.Fatal(err)
	}
	columns := []flyql.Column{flyql.NewColumn("level", flyql.TypeString)}
	diags := Diagnose(parsed, flyql.FromColumns(columns), nil)
	if len(diags) != 0 {
		t.Errorf("Expected 0 diagnostics, got %d", len(diags))
	}
}

func TestDiagnose_DottedColumnHighlightsBase(t *testing.T) {
	parsed, err := Parse("resource.service.name", Capabilities{Transformers: true})
	if err != nil {
		t.Fatal(err)
	}
	columns := []flyql.Column{flyql.NewColumn("level", flyql.TypeString)}
	diags := Diagnose(parsed, flyql.FromColumns(columns), nil)
	if len(diags) != 1 {
		t.Fatalf("Expected 1 diagnostic, got %d", len(diags))
	}
	if diags[0].Range.Start != 0 || diags[0].Range.End != 8 {
		t.Errorf("Range = {%d, %d}, want {0, 8} (just 'resource')", diags[0].Range.Start, diags[0].Range.End)
	}
}

// ---------------------------------------------------------------------------
// Diagnostic.Entry population (rich error objects in columns validator)
// ---------------------------------------------------------------------------

func TestColumnsDiagnoseEntryPopulated(t *testing.T) {
	parsed, err := Parse("foo", Capabilities{Transformers: true})
	if err != nil {
		t.Fatal(err)
	}
	cols := []flyql.Column{flyql.NewColumn("level", flyql.TypeString)}
	diags := Diagnose(parsed, flyql.FromColumns(cols), nil)
	if len(diags) != 1 {
		t.Fatalf("Expected 1 diagnostic, got %d", len(diags))
	}
	d := diags[0]
	if d.Entry.Name == "" {
		t.Errorf("Entry.Name empty for unknown_column")
	}
	codeStr, _ := d.Entry.Code.(string)
	if codeStr != d.Code {
		t.Errorf("Entry.Code = %v; want %q", d.Entry.Code, d.Code)
	}
}

type acceptsAnyT struct{}

func (acceptsAnyT) Name() string                                     { return "accepts_any" }
func (acceptsAnyT) Description() string                              { return "" }
func (acceptsAnyT) InputType() flyqltype.Type                        { return flyqltype.Any }
func (acceptsAnyT) OutputType() flyqltype.Type                       { return flyqltype.String }
func (acceptsAnyT) ArgSchema() []transformers.ArgSpec                { return nil }
func (acceptsAnyT) SQL(dialect, columnRef string, args []any) string { return columnRef }
func (acceptsAnyT) Apply(value any, args []any) any                  { return value }

type acceptsAnyReturningArrayT struct{}

func (acceptsAnyReturningArrayT) Name() string                                     { return "accepts_any_returning_array" }
func (acceptsAnyReturningArrayT) Description() string                              { return "" }
func (acceptsAnyReturningArrayT) InputType() flyqltype.Type                        { return flyqltype.Any }
func (acceptsAnyReturningArrayT) OutputType() flyqltype.Type                       { return flyqltype.Array }
func (acceptsAnyReturningArrayT) ArgSchema() []transformers.ArgSpec                { return nil }
func (acceptsAnyReturningArrayT) SQL(dialect, columnRef string, args []any) string { return columnRef }
func (acceptsAnyReturningArrayT) Apply(value any, args []any) any                  { return value }

func registryWithAny(t *testing.T) *transformers.TransformerRegistry {
	t.Helper()
	reg := transformers.DefaultRegistry()
	if err := reg.Register(acceptsAnyT{}); err != nil {
		t.Fatalf("Register(acceptsAnyT) failed: %v", err)
	}
	if err := reg.Register(acceptsAnyReturningArrayT{}); err != nil {
		t.Fatalf("Register(acceptsAnyReturningArrayT) failed: %v", err)
	}
	return reg
}

func TestColumnsValidator_AnyInputBypass(t *testing.T) {
	parsed, err := Parse("level|accepts_any", Capabilities{Transformers: true})
	if err != nil {
		t.Fatal(err)
	}
	cols := []flyql.Column{flyql.NewColumn("level", flyql.TypeInt)}
	diags := Diagnose(parsed, flyql.FromColumns(cols), registryWithAny(t))
	for _, d := range diags {
		if d.Code == flyql.CodeChainType {
			t.Errorf("expected no chain_type diagnostic, got %+v", d)
		}
	}
}

func TestColumnsValidator_ChainStrictAfterAnyInput(t *testing.T) {
	parsed, err := Parse("level|accepts_any_returning_array|upper", Capabilities{Transformers: true})
	if err != nil {
		t.Fatal(err)
	}
	cols := []flyql.Column{flyql.NewColumn("level", flyql.TypeInt)}
	diags := Diagnose(parsed, flyql.FromColumns(cols), registryWithAny(t))
	chainCount := 0
	for _, d := range diags {
		if d.Code == flyql.CodeChainType {
			chainCount++
		}
	}
	if chainCount != 1 {
		t.Fatalf("expected exactly 1 chain_type diagnostic, got %d (diags=%+v)", chainCount, diags)
	}
}

func TestColumnsParserErrorEntryPopulated(t *testing.T) {
	// Trigger a columns parser error: trailing pipe with no transformer.
	_, err := Parse("foo|", Capabilities{Transformers: true})
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	pe, ok := err.(*ParserError)
	if !ok {
		t.Fatalf("expected *ParserError, got %T", err)
	}
	if pe.Errno == 0 {
		t.Fatalf("expected non-zero Errno, got 0")
	}
	if pe.Entry.Name == "" {
		t.Errorf("Entry.Name empty for errno=%d", pe.Errno)
	}
	codeInt, _ := pe.Entry.Code.(int)
	if codeInt != pe.Errno {
		t.Errorf("Entry.Code = %v; want %d", pe.Entry.Code, pe.Errno)
	}
}

File: javascript/packages/flyql/test/columns/validator.test.js
import { describe, it, expect } from 'vitest'
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'
import { parse, diagnose } from '../../src/columns/index.js'
import { Column, ColumnSchema } from '../../src/core/column.js'
import {
    CODE_UNKNOWN_COLUMN,
    CODE_UNKNOWN_TRANSFORMER,
    CODE_CHAIN_TYPE,
    CODE_UNKNOWN_RENDERER,
    CODE_RENDERER_ARG_COUNT,
    CODE_RENDERER_ARG_TYPE,
    Diagnostic,
    ErrorEntry,
} from '../../src/core/validator.js'
import { VALIDATOR_REGISTRY } from '../../src/errors_generated.js'
import { Range } from '../../src/core/range.js'
import { Renderer, RendererRegistry, ArgSpec } from '../../src/renderers/index.js'
import { Type, parseFlyQLType } from '../../src/flyql_type.js'
import { Transformer, defaultRegistry } from '../../src/transformers/index.js'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const FIXTURE_PATH = path.join(__dirname, '..', '..', '..', '..', '..', 'tests-data', 'columns', 'validator.json')
const SHARED_DATA = JSON.parse(fs.readFileSync(FIXTURE_PATH, 'utf-8'))

function makeColumn(name, typeStr) {
    const t = typeStr ? parseFlyQLType(typeStr) : Type.Unknown
    return new Column(name, t, { matchName: name })
}

function buildColumnFromDef(d) {
    let children = null
    if (d.children && typeof d.children === 'object') {
        children = {}
        for (const [childName, childDef] of Object.entries(d.children)) {
            children[childName] = buildColumnFromDef(childDef)
        }
    }
    const t = d.type ? parseFlyQLType(d.type) : Type.Unknown
    return new Column(d.name, t, {
        matchName: d.name,
        children,
    })
}

function makeSchema(defs) {
    const m = {}
    for (const d of defs) {
        m[d.name] = buildColumnFromDef(d)
    }
    return new ColumnSchema(m)
}

function makeSchemaFromColumn(name, normalizedType) {
    return ColumnSchema.fromColumns([makeColumn(name, normalizedType)])
}

describe('Columns Validator', () => {
    describe('shared fixtures', () => {
        for (const tc of SHARED_DATA.tests) {
            it(tc.name, () => {
                const caps = tc.capabilities || { transformers: true }
                let parsedColumns
                try {
                    parsedColumns = parse(tc.input, caps)
                } catch {
                    parsedColumns = []
                }
                const schema = makeSchema(tc.columns)
                const diags = diagnose(parsedColumns, schema)
                expect(diags).toHaveLength(tc.expected_diagnostics.length)
                for (let i = 0; i < tc.expected_diagnostics.length; i++) {
                    const expected = tc.expected_diagnostics[i]
                    expect(diags[i].code).toBe(expected.code)
                    expect(diags[i].severity).toBe(expected.severity)
                    expect(diags[i].range.start).toBe(expected.range[0])
                    expect(diags[i].range.end).toBe(expected.range[1])
                    if (expected.message_contains) {
                        expect(diags[i].message).toContain(expected.message_contains)
                    }
                }
            })
        }
    })

    describe('additional tests', () => {
        it('empty array returns empty', () => {
            const diags = diagnose([], makeSchemaFromColumn('level', 'string'))
            expect(diags).toEqual([])
        })

        it('null returns empty', () => {
            const diags = diagnose(null, makeSchemaFromColumn('level', 'string'))
            expect(diags).toEqual([])
        })

        it('valid column with valid transformer returns empty', () => {
            const cols = parse('level|upper', { transformers: true })
            const diags = diagnose(cols, makeSchemaFromColumn('level', 'string'))
            expect(diags).toEqual([])
        })

        it('unknown column returns CODE_UNKNOWN_COLUMN', () => {
            const cols = parse('foo', { transformers: true })
            const diags = diagnose(cols, makeSchemaFromColumn('level', 'string'))
            expect(diags).toHaveLength(1)
            expect(diags[0].code).toBe(CODE_UNKNOWN_COLUMN)
            expect(diags[0].range.start).toBe(0)
            expect(diags[0].range.end).toBe(3)
        })

        it('unknown transformer returns CODE_UNKNOWN_TRANSFORMER', () => {
            const cols = parse('level|zzzz', { transformers: true })
            const diags = diagnose(cols, makeSchemaFromColumn('level', 'string'))
            expect(diags).toHaveLength(1)
            expect(diags[0].code).toBe(CODE_UNKNOWN_TRANSFORMER)
            expect(diags[0].range.start).toBe(6)
            expect(diags[0].range.end).toBe(10)
        })

        it('chain type mismatch returns CODE_CHAIN_TYPE', () => {
            // len outputs int, upper expects string
            const cols = parse('level|len|upper', { transformers: true })
            const diags = diagnose(cols, makeSchemaFromColumn('level', 'string'))
            expect(diags.some((d) => d.code === CODE_CHAIN_TYPE)).toBe(true)
        })

        class _AcceptsAny extends Transformer {
            get name() {
                return 'accepts_any'
            }
            get inputType() {
                return Type.Any
            }
            get outputType() {
                return Type.String
            }
            sql(dialect, columnRef) {
                return columnRef
            }
            apply(value) {
                return value
            }
        }

        class _AcceptsAnyReturningArray extends Transformer {
            get name() {
                return 'accepts_any_returning_array'
            }
            get inputType() {
                return Type.Any
            }
            get outputType() {
                return Type.Array
            }
            sql(dialect, columnRef) {
                return columnRef
            }
            apply(value) {
                return value
            }
        }

        function _registryWithAny() {
            const reg = defaultRegistry()
            reg.register(new _AcceptsAny())
            reg.register(new _AcceptsAnyReturningArray())
            return reg
        }

        it('Any-input transformer accepts any column type', () => {
            const cols = parse('level|accepts_any', { transformers: true })
            const diags = diagnose(cols, makeSchemaFromColumn('level', 'int'), _registryWithAny())
            expect(diags.some((d) => d.code === CODE_CHAIN_TYPE)).toBe(false)
        })

        it('chain remains strict after Any-input transformer returning array', () => {
            const cols = parse('level|accepts_any_returning_array|upper', { transformers: true })
            const diags = diagnose(cols, makeSchemaFromColumn('level', 'int'), _registryWithAny())
            const chainDiags = diags.filter((d) => d.code === CODE_CHAIN_TYPE)
            expect(chainDiags).toHaveLength(1)
            expect(chainDiags[0].message).toContain('upper')
        })

        it('multiple errors returns all diagnostics', () => {
            const cols = parse('foo, bar', { transformers: true })
            const diags = diagnose(cols, makeSchemaFromColumn('level', 'string'))
            expect(diags).toHaveLength(2)
            expect(diags[0].code).toBe(CODE_UNKNOWN_COLUMN)
            expect(diags[1].code).toBe(CODE_UNKNOWN_COLUMN)
        })

        it('dotted column highlights only base segment', () => {
            const cols = parse('resource.service.name', { transformers: true })
            const diags = diagnose(cols, makeSchemaFromColumn('level', 'string'))
            expect(diags).toHaveLength(1)
            expect(diags[0].range.start).toBe(0)
            expect(diags[0].range.end).toBe(8) // "resource"
        })
    })

    describe('renderers', () => {
        const caps = { transformers: true, renderers: true }

        class HrefRenderer extends Renderer {
            get name() {
                return 'href'
            }
            get argSchema() {
                return [new ArgSpec(Type.String, true)]
            }
        }

        class TruncateRenderer extends Renderer {
            get name() {
                return 'truncate'
            }
            get argSchema() {
                return [new ArgSpec(Type.Int, true)]
            }
        }

        function makeURLSchema() {
            return makeSchemaFromColumn('url', 'string')
        }

        function registry() {
            const reg = new RendererRegistry()
            reg.register(new HrefRenderer())
            reg.register(new TruncateRenderer())
            return reg
        }

        it('unknown renderer returns CODE_UNKNOWN_RENDERER', () => {
            const cols = parse('url as link|unknown("x")', caps)
            const diags = diagnose(cols, makeURLSchema(), null, registry())
            expect(diags).toHaveLength(1)
            expect(diags[0].code).toBe(CODE_UNKNOWN_RENDERER)
        })

        it('valid renderer returns zero diagnostics', () => {
            const cols = parse('url as link|href("/x")', caps)
            const diags = diagnose(cols, makeURLSchema(), null, registry())
            expect(diags).toHaveLength(0)
        })

        it('renderer arg count mismatch returns CODE_RENDERER_ARG_COUNT', () => {
            const cols = parse('url as link|href("/x", "extra")', caps)
            const diags = diagnose(cols, makeURLSchema(), null, registry())
            expect(diags.some((d) => d.code === CODE_RENDERER_ARG_COUNT)).toBe(true)
        })

        it('renderer arg type mismatch returns CODE_RENDERER_ARG_TYPE', () => {
            const cols = parse('url as link|href(42)', caps)
            const diags = diagnose(cols, makeURLSchema(), null, registry())
            expect(diags.some((d) => d.code === CODE_RENDERER_ARG_TYPE)).toBe(true)
        })

        it('per-renderer diagnose hook appends custom diagnostics', () => {
            class HrefWithHook extends Renderer {
                get name() {
                    return 'href'
                }
                get argSchema() {
                    return [new ArgSpec(Type.String, true)]
                }
                diagnose(args, _col) {
                    if (args && args[0] && !args[0].includes('{{value}}')) {
                        return [new Diagnostic(new Range(0, 1), 'href missing placeholder', 'warning', 'custom_href')]
                    }
                    return []
                }
            }
            const reg = new RendererRegistry()
            reg.register(new HrefWithHook())
            const cols = parse('url as link|href("/static")', caps)
            const diags = diagnose(cols, makeURLSchema(), null, reg)
            expect(diags.some((d) => d.code === 'custom_href')).toBe(true)
        })

        it('registry-level chain diagnose hook appends diagnostics', () => {
            const reg = registry()
            reg.setDiagnose((col, chain) => {
                const names = chain.map((r) => r.name)
                const ti = names.indexOf('truncate')
                const hi = names.indexOf('href')
                if (ti >= 0 && hi >= 0 && ti < hi) {
                    return [new Diagnostic(new Range(0, 1), 'href cannot follow truncate', 'error', 'chain_forbidden')]
                }
                return []
            })
            const cols = parse('url as link|truncate(10)|href("/x")', caps)
            const diags = diagnose(cols, makeURLSchema(), null, reg)
            expect(diags.some((d) => d.code === 'chain_forbidden')).toBe(true)
        })

        it('null registry emits unknown_renderer for every renderer', () => {
            const cols = parse('url as link|href("/x")', caps)
            const diags = diagnose(cols, makeURLSchema(), null, null)
            expect(diags).toHaveLength(1)
            expect(diags[0].code).toBe(CODE_UNKNOWN_RENDERER)
        })
    })

    describe('Diagnostic.error population', () => {
        function _hrefRegistry() {
            const reg = new RendererRegistry()
            class _Href extends Renderer {
                get name() {
                    return 'href'
                }
                get argSchema() {
                    return [new ArgSpec(Type.String, true)]
                }
            }
            reg.register(new _Href())
            return reg
        }

        const baseSchema = ColumnSchema.fromColumns([makeColumn('url', 'string')])
        const baseCaps = { transformers: true, renderers: true }

        const cases = [
            {
                query: 'foo',
                schema: ColumnSchema.fromColumns([makeColumn('level', 'string')]),
                code: CODE_UNKNOWN_COLUMN,
                withRenderer: false,
            },
            {
                query: 'level|zzzz',
                schema: ColumnSchema.fromColumns([makeColumn('level', 'string')]),
                code: CODE_UNKNOWN_TRANSFORMER,
                withRenderer: false,
            },
            {
                query: 'level|len|upper',
                schema: ColumnSchema.fromColumns([makeColumn('level', 'string')]),
                code: CODE_CHAIN_TYPE,
                withRenderer: false,
            },
            { query: 'url as link|wat("/x")', schema: baseSchema, code: CODE_UNKNOWN_RENDERER, withRenderer: true },
            { query: 'url as link|href', schema: baseSchema, code: CODE_RENDERER_ARG_COUNT, withRenderer: true },
            { query: 'url as link|href(123)', schema: baseSchema, code: CODE_RENDERER_ARG_TYPE, withRenderer: true },
        ]

        it.each(cases)('populates error for $code', ({ query, schema, code, withRenderer }) => {
            const caps = withRenderer ? baseCaps : { transformers: true }
            const cols = parse(query, caps)
            const reg = withRenderer ? _hrefRegistry() : null
            const diags = diagnose(cols, schema, null, reg)
            const matching = diags.filter((d) => d.code === code)
            expect(matching.length, `no diag with code ${code}`).toBeGreaterThan(0)
            for (const d of matching) {
                expect(d.error).not.toBeNull()
                expect(d.error).toBeInstanceOf(ErrorEntry)
                expect(d.error.code).toBe(d.code)
                expect(d.error.name).toBe(VALIDATOR_REGISTRY[d.code].name)
            }
        })

        it('user-extension Diagnostic with custom code has error=null', () => {
            const d = new Diagnostic(new Range(0, 1), 'msg', 'warning', 'custom_href_no_placeholder')
            expect(d.error).toBeNull()
        })
    })
})
